### PR TITLE
[lldb][test] Remove expectedFailureIfFn

### PR DESCRIPTION
**Pull request by rupprecht**
Switching to modern `unittest` in 5b386158aacac4b41126983a5379d36ed413d0ea needs xfail annotations to be known prior to test running. In contrast, skipping can happen at any time, even during test execution.

Thus, `expectedFailureIfFn` inherently doesn't work. Either we eagerly evaluate the function and use `expectedFailureIf` instead, or we use a skip annotation to lazily evaluate the function and potentially skip the test right before it starts.

- For `expectedFailureAndroid`, the intent seems to be that certain tests _should_ work on android, but don't. Thus, xfail is appropriate, to ensure the test is re-enabled once those bugs are ever fixed.
- For the other uses in individual tests, those generally seem to be cases where the test environment doesn't support the setup required by the test, and so it isn't meaningful to run the test at all. For those, a drop-in replacement to `skipTestIfFn` works.

### CODE REVIEW

**Review by DavidSpickett — APPROVED** [review]
LGTM.

Please hold off on merging until the Arm lldb bots are working again. I've landed the fix so it shouldn't be too long now.

**Comment by rupprecht** [comment]
The arm bots still haven't picked up the change :(

Green Dragon is back online, and TestRequireHWBreakpoints.py is failing: https://green.lab.llvm.org/green/view/LLDB/job/as-lldb-cmake/16077/testReport/lldb-api/functionalities_breakpoint_hardware_breakpoints_require_hw_breakpoints/TestRequireHWBreakpoints_py/. This PR should fix that.

It would be nice to wait for the arm bots to be back in a normal state, but I also don't want to leave green dragon red for too long, so I should probably land this today even if the other arm bots aren't reloaded. I'm not sure what the buildbot process is; @gkistanova, do you have to manually restart something to pick up config changes? e.g. https://lab.llvm.org/buildbot/#/builders/96/builds/52991 is still running `ninja "" ""`.

**Comment by JDevlieghere** [comment]
Yeah given GreenDragon has been red for 20h, I'd like to get this merged sooner rather than later. @DavidSpickett what's the benefit of waiting for the arm bots to come back up?

**Comment by rupprecht** [comment]
> Yeah given GreenDragon has been red for 20h, I'd like to get this merged sooner rather than later. @DavidSpickett what's the benefit of waiting for the arm bots to come back up?

Broadly speaking, if a buildbot is down, it interferes with bisection if something unexpected happens, particularly for commits like this and 5b386158aacac4b41126983a5379d36ed413d0ea which have a broader impact. So if the buildbot comes back up and TestRequireHWBreakpoints.py is failing, it'd be useful to see if it was because of this commit or something else.

... that's what I figure at least. In hindsight it would have been a good idea to check that buildbots were mostly green before landing the original change. I just assumed things were always green :)

I feel relatively confident that this should fix the green dragon builds, and is worth the risk that it may create more of a mess in arm bots (but hopefully not!). Reverting 5b386158aacac4b41126983a5379d36ed413d0ea is also an option, but it seems like it has already exposed some crashes in `lldb-dap` that were previously ignored, so I think it's worth keeping. I'll merge this shortly.

**Comment by DavidSpickett** [comment]
> @DavidSpickett what's the benefit of waiting for the arm bots to come back up?

If you give me the choice of more potential chaos or less, I'm always going to choose less :)

But yes this shouldn't be too difficult to pick out if it does fail.

**Comment by rastogishubham** [comment]
Hi @rupprecht,

It seems like this patch has still not fixed the issue in greendragon and the test is still flaky

https://green.lab.llvm.org/green/view/LLDB/job/lldb-cmake/
https://green.lab.llvm.org/green/view/LLDB/job/lldb-cmake-matrix/
https://green.lab.llvm.org/green/view/LLDB/job/lldb-cmake-sanitized/

@JDevlieghere Do you have any idea as to why that is still happening?

**Comment by rupprecht** [comment]
I believe the old unittest framework may have been ignoring failures that happen during teardown.

I have started looking at https://github.com/llvm/llvm-project/issues/81686 a bit, but +1 to suppressing this temporarily -- I should be able to land that in a bit.

**Comment by rupprecht** [comment]
Skipped the test in 65c25a4c40865a0e460c0fecb1b33f7cf7455573

Tear down behavior was patched for LLDB back in b1490b6172c48cf802b24bd738c90eeb019436fe. I'm guessing that explains why we weren't noticing those errors -- they get logged as cleanup_errors, not as errors. (Maybe I'm missing something). By switching to standard unittest and dropping that patch, we're going to catch more tear down issues.

**Pull request merged**
